### PR TITLE
ARM Template for Azure Batch

### PR DESCRIPTION
**Pull request by lawrencegripper**
This template creates an auto-scaling Azure Batch Account and node pool which support Container based tasks. 

When there is no work in the queue the pool scales down to 0 machines, this reduces costs but does mean tasks will have a start time of around ~5/8 mins depending on when the auto-scaling rule was last evaluated. 

To work around a limitation in the ARM API the pool creation is handled by an ACI container with a MSI assigned to it. Hopefully this can be removed in future. 

Remaining work:

- [x] Integrate with the top level template
- [x] Run end-2-end test a couple more times

### CODE REVIEW

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json [hunk 1]
```diff
@@ -0,0 +1,199 @@
...
+            "defaultValue": 3
+        },
+        "virtualMachineSize": {
+            "type": "string",
+            "defaultValue": "Standard_F8s_v2",
+            "metadata": {
+                "description": "Size of VMs in the VM Scale Set."
+            }
+        },
+        "storageAccountsku": {
```
Comment (jjcollinge) [inline opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: nit: SKU should be camel cased

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json [hunk 2]
```diff
@@ -0,0 +1,199 @@
...
+                "scope": "[resourceGroup().id]"
+            }
+        },
+        {
+            "comments": "This is the storage account used by Azure Batch for file processing/storage",
+            "type": "Microsoft.Storage/storageAccounts",
+            "name": "[variables('storageAccountname')]",
+            "apiVersion": "2016-01-01",
+            "location": "[parameters('location')]",
+            "comments": "This storage account is used to associate to a batch account",
```
Comment (jjcollinge) [inline opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: duplicate comment?

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json [hunk 3]
```diff
@@ -0,0 +1,199 @@
...
+              "restartPolicy": "Never",
+              "containers": [
+                {
+                  "name": "azure-cli",
+                  "properties": {
+                    "image": "microsoft/azure-cli",
+                    "command": [
+                      "/bin/bash",
+                      "-c",
+                      "[concat('az login --identity -u ', resourceId('Microsoft.ManagedIdentity/userAssignedIdentities', variables('identityName')), ' && az batch account login --name ', variables('batchAccountName'), ' -g ', resourceGroup().name, ' && curl -fsSL ', parameters('_artifactsLocation'), '/azurebatch/configurepool.py', parameters('_artifactsLocationSasToken'), ' > configurepool.py && python3 ./configurepool.py \"', parameters('poolId'), '\" ', parameters('virtualMachineSize'),  ' \"', parameters('mountArgs'),  '\" ', parameters('_artifactsLocation'),  ' ', parameters('_artifactsLocationSasToken'),  ' ', parameters('virtualNetworkSubnetId'))]"
```
Comment (jjcollinge) [inline opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: could the script not login etc. to stop logic being here and in the script?

File: opencga-app/app/scripts/azure/arm/azurebatch/pool.json [hunk 4]
```diff
@@ -0,0 +1,48 @@
+{
+    "id": "POOL_ID_HERE",
+    "vmSize": "VM_SIZE_HERE",
+    "enableAutoScale": true,
+    "autoScaleFormula": "startingNumberOfVMs = 1;maxNumberofVMs = 3;pendingTaskSamplePercent = $PendingTasks.GetSamplePercent(160 * TimeInterval_Second);pendingTaskSamples = pendingTaskSamplePercent < 70 ? startingNumberOfVMs : avg($PendingTasks.GetSample(160 * TimeInterval_Second));$TargetDedicatedNodes=min(maxNumberofVMs, pendingTaskSamples);",
```
Comment (jjcollinge) [inline opencga-app/app/scripts/azure/arm/azurebatch/pool.json]: Could this formula be parameterized in the script if needed? Let the user set the scale limits.

Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/azurebatch/pool.json]: Good spot, that was something that got lost in my move from ARM -> ACI ... will fix that up as have a param for `maxNodeCount`

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json (hunk 3 above)
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: Yup I think that's a good call too, will make this line less horrible to read

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json (hunk 2 above)
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: woops ta

File: opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json (hunk 1 above)
Comment (lawrencegripper) [reply opencga-app/app/scripts/azure/arm/azurebatch/azuredeploy.json]: thanks fixed

**Comment by lawrencegripper** [comment]
Thanks for the review @jjcollinge I've fixed things up and will move onto integrating it into the main template tomorrow ready for merging :smile:

**Comment by lawrencegripper** [comment]
@jjcollinge and @marrobi What do you think I do here? When I create the template for the first time you sometimes get this error:

```json
{
  "status": "Failed",
  "error": {
    "code": "ResourceDeploymentFailure",
    "message": "The resource operation completed with terminal provisioning state 'Failed'.",
    "details": [
      {
        "code": "DeploymentFailed",
        "message": "At least one resource deployment operation failed. Please list deployment operations for details. Please see https://aka.ms/arm-debug for usage details.",
        "details": [
          {
            "code": "BadRequest",
            "message": "{\r\n  \"error\": {\r\n    \"code\": \"PrincipalNotFound\",\r\n    \"message\": \"Principal 18a829a8c947440287e14445d35994ae does not exist in the directory 72f988bf-86f1-41af-91ab-2d7cd011db47.\"\r\n  }\r\n}"
          }
        ]
      }
    ]
  }
}
```

It looks like this is a replication issue with the new `Identity` not being available, rerunning the template resolves the issue but I can't see any way I can work around this my end other than to say .... if this happens run it twice ... Got any ideas?

**Comment by lawrencegripper** [comment]
So I've worked around the above issue by making the `RoleAssignment` depend on another resource which slows it down and appears to prevent the `PrincipalNotFound` error. 

When the VMs come up in the AzureBatch pool then seem to already have some `apt install` stuff in progress which meant the `mount.py` script failed. I've moved this over now to retry doing it's installs for the utils required to mount the shares and that seems to now behave: 

```
Attempting, with retries, to: install cifs-utils
Attempt #1
Attempt to install cifs-utils
Failed install cifs-utils error: Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
Failed:Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
retrying in 3secs
Attempt #2
Attempt to install cifs-utils
Failed install cifs-utils error: Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
Failed:Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
retrying in 6secs
Attempt #3
Attempt to install cifs-utils
Failed install cifs-utils error: Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
Failed:Command '['apt', 'install', 'cifs-utils', '-y']' returned non-zero exit status 100
retrying in 9secs
Attempt #4
Attempt to install cifs-utils
Install completed successfully
Succeeded to: install cifs-utils after 4 retries
Mounting primary
Done editing fstab ... attempting mount
Attempting, with retries, to: mount shares
Attempt #1
Succeeded to: mount shares after 1 retries
```

**Comment by lawrencegripper** [comment]
So based on some conversations I've had it looks like the workaround using ACI to create the pool may no longer be needed in a week or two. Do we want to go ahead and merge this and then revisit to simplify or hold off? @marrobi 

My vote is to merge and then revisit.

**Pull request merged**